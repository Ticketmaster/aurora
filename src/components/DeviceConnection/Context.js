import React from "react";

export const { Provider, Consumer } = React.createContext({ "4g": true });
