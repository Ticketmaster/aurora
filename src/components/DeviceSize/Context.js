import React from "react";

export const { Provider, Consumer } = React.createContext({ isSmall: true });

export const withDeviceSize = Component => props => (
  <Consumer>{value => <Component deviceSize={value} {...props} />}</Consumer>
);
