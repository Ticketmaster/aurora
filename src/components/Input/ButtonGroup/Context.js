import React from "react";

export const { Provider, Consumer } = React.createContext({
  selected: null,
  onChange: null
});

export const withButtonGroupConsumer = Component => props => (
  <Consumer>
    {({ onChange: onGroupChange, value: groupValue = {} }) => (
      <Component
        onGroupChange={onGroupChange}
        groupValue={groupValue}
        {...props}
      />
    )}
  </Consumer>
);
