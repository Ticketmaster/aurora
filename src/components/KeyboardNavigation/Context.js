import React from "react";

export const { Provider, Consumer } = React.createContext({
  focused: null,
  focusSelected: () => null
});
