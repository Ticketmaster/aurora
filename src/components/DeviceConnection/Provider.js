import React, { Component } from "react";
import PropTypes from "prop-types";

import { Provider } from "./Context";

import { deviceConnectionStateShape } from "./shape";
import { FALSY_CONNECTIONS_STATE } from "./constants";

class DeviceConnectionProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    defaultState: PropTypes.shape(deviceConnectionStateShape)
  };

  static defaultProps = {
    defaultState: false
  };

  constructor(props) {
    super(props);
    const { defaultState } = props;

    this.state = defaultState || this.initialState;
  }

  componentDidMount() {
    const { effectiveType, type } = global.navigator.connection;

    if (effectiveType || type) {
      this.subscribe();
      this.update({ [effectiveType || type]: true, isInitialized: true });
    }
  }

  componentWillUnmount = () => {
    const { isInitialized } = this.state;

    if (isInitialized) {
      this.unsubscribe();
    }
  };

  onConnectionChange() {
    const { effectiveType, type } = global.navigator.connection;
    this.update({ [effectiveType || type]: true });
  }

  initialState = {
    ...FALSY_CONNECTIONS_STATE,
    "4g": true,
    isInitialized: false
  };

  subscribe() {
    global.navigator.connection.addEventListener(
      "change",
      this.onConnectionChange
    );
  }

  // eslint-disable-next-line
  unsubscribe() {
    global.navigator.connection.removeEventListener("change");
  }

  update(payload) {
    this.setState({
      ...FALSY_CONNECTIONS_STATE,
      ...payload
    });
  }

  render() {
    const { children } = this.props;
    return <Provider value={this.state}>{children}</Provider>;
  }
}

export default DeviceConnectionProvider;
