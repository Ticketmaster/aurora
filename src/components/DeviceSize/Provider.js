/* global window */
import React from "react";
import PropTypes from "prop-types";

import { Provider } from "./Context";
import constants from "../../theme/constants";

export default class DeviceSizeProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    fallbackDetection: PropTypes.func,
    cssOnly: PropTypes.bool
  };

  static defaultProps = { fallbackDetection: null, cssOnly: false };

  initialState = { isSmall: true }; // eslint-disable-line

  state = this.initialState;

  componentDidMount() {
    if (
      typeof window === "undefined" ||
      typeof window.matchMedia !== "function"
    )
      return;

    if (this.props.cssOnly) {
      this.setSize();
      return;
    }

    this.smallMedia = window.matchMedia(constants.breakpoints.small);
    this.mediumMedia = window.matchMedia(constants.breakpoints.medium);
    this.largeMedia = window.matchMedia(constants.breakpoints.large);
    this.xLargeMedia = window.matchMedia(constants.breakpoints.xLarge);
    this.smallMedia.addListener(this.setSize);
    this.mediumMedia.addListener(this.setSize);
    this.largeMedia.addListener(this.setSize);
    this.xLargeMedia.addListener(this.setSize);
    this.setSize();
  }

  componentDidUpdate() {
    if (!this.props.cssOnly) return;

    this.unsubscribe();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  setSize = () => {
    if (this.props.cssOnly) {
      this.setState(() => ({ cssOnly: true }));
      return;
    }

    this.setState(() => ({
      isSmall: this.smallMedia.matches,
      isMedium: this.mediumMedia.matches && !this.largeMedia.matches,
      isLarge: this.largeMedia.matches && !this.xLargeMedia.matches,
      isXLarge: this.xLargeMedia.matches,
      cssOnly: false
    }));
  };

  unsubscribe = () => {
    if (this.smallMedia) this.smallMedia.removeListener(this.setSize);
    if (this.mediumMedia) this.mediumMedia.removeListener(this.setSize);
    if (this.largeMedia) this.largeMedia.removeListener(this.setSize);
    if (this.xLargeMedia) this.xLargeMedia.removeListener(this.setSize);
  };

  render() {
    const { fallbackDetection } = this.props;
    const val = fallbackDetection ? fallbackDetection() : this.state;
    return (
      <Provider value={val || this.initialState}>
        {this.props.children}
      </Provider>
    );
  }
}
