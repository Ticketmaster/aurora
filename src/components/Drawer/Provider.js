/* global window: true */
/* eslint react/sort-comp: off, react/no-unused-state: off */
import React, { Children } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { Provider } from "./Context";
import { ContentWrapper } from "./Drawer.styles";
import Portal from "../Portal";

export default class DrawerProvider extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    className: PropTypes.string
  };

  static defaultProps = { className: "" };

  contentContainer = React.createRef();

  scrollPosition = 0;

  shouldComponentUpdate(prevProps, prevState) {
    return (
      prevState.isOpen !== this.state.isOpen ||
      prevState.content !== this.state.content ||
      prevProps.children !== this.props.children
    );
  }

  componentDidUpdate() {
    const { isOpen } = this.state;
    /* istanbul ignore next */
    if (
      !isOpen &&
      typeof window !== "undefined" &&
      typeof window.scrollTo === "function"
    ) {
      window.scrollTo(0, this.scrollPosition);
    }
  }

  setContent = content => this.setState(() => ({ content }));

  toggleDrawer = () => {
    this.setState(({ isOpen }) => {
      /* istanbul ignore else */
      if (typeof window !== "undefined" && !isOpen) {
        this.scrollPosition = window.pageYOffset;
      }
      return { isOpen: !isOpen };
    });
  };

  state = {
    isOpen: false,
    toggleDrawer: this.toggleDrawer,
    setContent: this.setContent,
    content: null
  };

  render() {
    const { className, children } = this.props;
    const { isOpen, content } = this.state;
    return (
      <Provider value={this.state}>
        <ContentWrapper
          className={classNames({ "drawer--open": isOpen }, className)}
          innerRef={this.contentContainer}
        >
          {Children.only(children)}
          {content && <Portal>{content}</Portal>}
        </ContentWrapper>
      </Provider>
    );
  }
}
