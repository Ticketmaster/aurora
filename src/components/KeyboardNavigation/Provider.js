import React from "react";
import PropTypes from "prop-types";
import { Provider } from "./Context";
import { ARROWUP, ARROWDOWN, SPACEBAR } from "../../utils/keyCharCodes";

const { forEach } = React.Children;

const getSelectedChild = (children, selectedValue) =>
  React.Children.toArray(children).find(
    thisArg => thisArg.props.value === selectedValue
  );

export default class KeyBoardProvider extends React.Component {
  static propTypes = {
    children(props, propName, componentName) {
      if (props[propName] === null) return new Error("Invalid Child");
      forEach(props[propName], child => {
        if (!child.props.value || !child.props.index) {
          return new Error(
            `Invalid Child \`${child}\` supplied to` +
              ` \`${componentName}\`. Validation failed.`
          );
        }
        return null;
      });
      return null;
    },

    role: PropTypes.string.isRequired,
    className: PropTypes.string,
    keywordSearch: PropTypes.bool,
    keyBoardRef: PropTypes.func
  };

  static defaultProps = {
    children: null,
    keyBoardRef: null,
    className: null,
    keywordSearch: false
  };

  static getDerivedStateFromProps(props, state) {
    const { selected, children } = props;

    if (!selected || !selected.length) return null;

    const selectedValue = selected[selected.length - 1];

    if (state.firstUpdate === true) {
      const selectedChild = getSelectedChild(children, selectedValue);

      return {
        focused: selectedChild ? selectedChild.props.index : 0,
        firstUpdate: false,
        prevSelectedValue: selectedValue
      };
    }

    if (selectedValue !== state.prevSelectedValue) {
      const selectedChild = getSelectedChild(children, selectedValue);

      return {
        focused: selectedChild ? selectedChild.props.index : 0,
        prevSelectedValue: selectedValue
      };
    }

    return null;
  }

  onClick = ({ index }) => {
    this.setState(() => ({ focused: index }));
    this.setState({ word: "" });
  };

  onKeyDown = event => {
    const { keywordSearch } = this.props;

    // heres the logic for looking up words with in the drop down options with the keyboard
    // it currently matches based on the text that has been typed
    if (keywordSearch) {
      const { word } = this.state;
      if (event.keyCode > 48 && event.keyCode < 90) {
        const temp = word + event.key;
        this.setState({ word: temp }, () => this.searchValue(temp));
      }

      if (event.keyCode === SPACEBAR) this.setState({ word: "" });
    }

    if ([ARROWUP, ARROWDOWN].includes(event.keyCode)) {
      event.preventDefault();
      this.cycleSelect(event.keyCode);
    }
  };

  searchValue = string => {
    const { children } = this.props;

    const childrenArray = React.Children.toArray(children);

    const firstMatch = childrenArray.find(
      thisArg =>
        thisArg.props.children.substring(0, string.length).toLowerCase() ===
        string
    );
    if (firstMatch) this.setState(() => ({ focused: firstMatch.props.index }));
  };

  cycleSelect = key => {
    const { focused } = this.state;
    const { children } = this.props;

    if (key === ARROWUP && focused <= 0) {
      this.setState(() => ({ focused: children.length - 1 }));
    }

    if (key === ARROWUP && focused > 0) {
      this.setState(prevState => ({ focused: prevState.focused - 1 }));
    }

    if (key === ARROWDOWN && focused >= children.length - 1) {
      this.setState(() => ({ focused: 0 }));
    }

    if (key === ARROWDOWN && focused < children.length - 1) {
      this.setState(prevState => ({ focused: prevState.focused + 1 }));
    }
  };

  /* eslint-disable */
  state = {
    focused: 0,
    focusSelected: this.onClick,
    word: "",
    firstUpdate: true,
    prevSelectedValue: ""
  };
  /* eslint-enable */

  render() {
    const { role, className, keyBoardRef } = this.props;
    return (
      <Provider value={this.state}>
        {/* eslint-disable */}
        <div
          role={role}
          className={className}
          ref={keyBoardRef}
          onKeyDown={this.onKeyDown}
        >
          {this.props.children}
        </div>
        {/* eslint-enable */}
      </Provider>
    );
  }
}
