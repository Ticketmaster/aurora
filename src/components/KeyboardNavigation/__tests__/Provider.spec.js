import React from "react";
import { renderIntoDocument, cleanup, fireEvent } from "react-testing-library";

import Provider from "../Provider";
import { Consumer } from "../Context";

const TesingConsumers = (
  { children, index, datatestid } // eslint-disable-line
) => (
  <Consumer>
    {({ focused }) => (
      <div data-isfocues={focused === index} data-testid={datatestid}>
        {children}
      </div>
    )}
  </Consumer>
);

const createTestElement = () => (
  <Provider role="option">
    <TesingConsumers datatestid="button0" index={0}>
      Testing1
    </TesingConsumers>
    <TesingConsumers datatestid="button1" index={1}>
      Testing2
    </TesingConsumers>
    <TesingConsumers datatestid="button2" index={2}>
      Testing3
    </TesingConsumers>
    <TesingConsumers datatestid="button3" index={3}>
      Testing4
    </TesingConsumers>
  </Provider>
);

const createTestElementSelected = (selected = []) => (
  <Provider selected={selected} role="option">
    <TesingConsumers value="button0" datatestid="button0" index={0}>
      Testing1
    </TesingConsumers>
    <TesingConsumers value="button1" datatestid="button1" index={1}>
      Testing2
    </TesingConsumers>
    <TesingConsumers value="button2" datatestid="button2" index={2}>
      Testing3
    </TesingConsumers>
    <TesingConsumers value="button3" datatestid="button3" index={3}>
      Testing4
    </TesingConsumers>
  </Provider>
);

describe("KeyBoardProvider", () => {
  afterEach(cleanup);

  it("Up outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(createTestElement());

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Up within range of number of Children", () => {
    const { getByTestId, container } = renderIntoDocument(createTestElement());

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 38 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Down within range of number of Children", () => {
    const { getByTestId, container } = renderIntoDocument(createTestElement());

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Down outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(createTestElement());

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });
    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Down outside range of number of Children ", () => {
    const { getByTestId, container } = renderIntoDocument(createTestElement());

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 3 });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Change focus position when 'selected' prop was changed from outside ", () => {
    const { getByTestId, container, rerender } = renderIntoDocument(
      createTestElementSelected(["button0"])
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    rerender(createTestElementSelected(["button3"]));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Set default focus position when 'selected' prop is invalid on the first update ", () => {
    const { container } = renderIntoDocument(
      createTestElementSelected(["nonexistent-value"])
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Reset focus position when 'selected' prop was changed from outside ", () => {
    const { getByTestId, container, rerender } = renderIntoDocument(
      createTestElementSelected(["button0"])
    );

    fireEvent.keyDown(getByTestId("button0"), { keyCode: 40 });

    rerender(createTestElementSelected(["nonexistent-value"]));

    expect(container.firstChild).toMatchSnapshot();
  });
});
