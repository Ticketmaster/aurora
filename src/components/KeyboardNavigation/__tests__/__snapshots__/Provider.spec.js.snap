// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyBoardProvider Change focus position when 'selected' prop was changed from outside  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="true"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Down outside range of number of Children  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="true"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Down outside range of number of Children  2`] = `
<div
  role="option"
>
  <div
    data-isfocues="true"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Down within range of number of Children 1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="true"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Reset focus position when 'selected' prop was changed from outside  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="true"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Set default focus position when 'selected' prop is invalid on the first update  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="true"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Up outside range of number of Children  1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="false"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="true"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;

exports[`KeyBoardProvider Up within range of number of Children 1`] = `
<div
  role="option"
>
  <div
    data-isfocues="false"
    data-testid="button0"
  >
    Testing1
  </div>
  <div
    data-isfocues="false"
    data-testid="button1"
  >
    Testing2
  </div>
  <div
    data-isfocues="true"
    data-testid="button2"
  >
    Testing3
  </div>
  <div
    data-isfocues="false"
    data-testid="button3"
  >
    Testing4
  </div>
</div>
`;
